File: src/page/info/StageTable.java
package page.info;

import common.CommonStatic;
import common.util.Data;
import common.util.stage.SCDef.Line;
import common.util.stage.SCGroup;
import common.util.stage.Stage;
import common.util.unit.Enemy;
import page.MainFrame;
import page.Page;
import page.support.AbJTable;
import page.support.EnemyTCR;

import java.awt.*;
import java.awt.event.MouseEvent;

public class StageTable extends AbJTable {

    private static final long serialVersionUID = 1L;

    private static String[] title;

    static {
        redefine();
    }

    public static void redefine() {
        title = Page.get(1, "t", 9);
    }

    protected Object[][] data;

    private final Page page;

    protected StageTable(Page p) {
        page = p;

        setDefaultRenderer(Enemy.class, new EnemyTCR());
    }

    @Override
    public Class<?> getColumnClass(int c) {
        if (c == 1)
            return Enemy.class;
        else
            return Object.class;
    }

    @Override
    public int getColumnCount() {
        return title.length;
    }

    @Override
    public String getColumnName(int arg0) {
        return title[arg0];
    }

    @Override
    public int getRowCount() {
        if (data == null)
            return 0;
        return data.length;
    }

    @Override
    public String getToolTipText(MouseEvent e) {
        if (columnAtPoint(e.getPoint()) == 2) {
            return "{hp, atk}";
        } else {
            return null;
        }
    }

    @Override
    public Object getValueAt(int r, int c) {
        if (data == null || r < 0 || c < 0 || r >= data.length || c >= data[r].length)
            return null;
        if (c == 2)
            return data[r][c] + "%";
        return data[r][c];
    }

    protected void clicked(Point p) {
        if (data == null)
            return;
        int c = getColumnModel().getColumnIndexAtX(p.x);
        c = lnk[c];
        int r = p.y / getRowHeight();
        if (r < 0 || r >= data.length || c != 1)
            return;
        Enemy e = (Enemy) data[r][c];

        if (!(data[r][2] instanceof String))
            return;

        int[] d = CommonStatic.parseIntsN((String) data[r][2]);

        MainFrame.changePanel(new EnemyInfoPage(page, e, d[0], d[1]));
    }

    protected void setData(Stage st) {
        Line[] info = st.data.getSimple();
        data = new Object[info.length][9];
        for (int i = 0; i < info.length; i++) {
            int ind = info.length - i - 1;
            data[ind][1] = info[i].enemy.get();
            data[ind][0] = info[i].boss == 1 ? "boss" : "";
            data[ind][2] = CommonStatic.toArrayFormat(info[i].multiple, info[i].mult_atk);
            data[ind][3] = info[i].number == 0 ? "infinite" : info[i].number;
            if (info[i].castle_0 >= info[i].castle_1)
                data[ind][4] = info[i].castle_0 + "%";
            else
                data[ind][4] = info[i].castle_0 + "~" + info[i].castle_1 + "%";
            if (Math.abs(info[i].spawn_0) >= Math.abs(info[i].spawn_1))
                data[ind][5] = info[i].spawn_0;
            else
                data[ind][5] = info[i].spawn_0 + "~" + info[i].spawn_1;

            if (info[i].respawn_0 == info[i].respawn_1)
                data[ind][6] = info[i].respawn_0;
            else
                data[ind][6] = info[i].respawn_0 + "~" + info[i].respawn_1;

            data[ind][7] = info[i].layer_0 == info[i].layer_1 ? info[i].layer_0
                    : info[i].layer_0 + "~" + info[i].layer_1;
            int g = info[i].group;
            SCGroup scg = st.data.sub.get(g);
            data[ind][8] = scg == null ? g != 0 ? Data.trio(g) + " - invalid" : "" : scg.toString();

        }
    }

}

File: src/page/info/StageTable.kt
package page.info

import common.CommonStatic
import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonEncoder
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.util.Data
import common.util.stage.EStage
import common.util.stage.SCGroup
import common.util.stage.Stage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.Enemy
import common.util.unit.UnitLevel
import io.BCPlayer
import page.JL
import page.MainFrame
import page.Page
import page.anim.AnimBox
import page.support.AbJTable
import page.support.EnemyTCR
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter
import java.awt.Point
import java.awt.event.MouseEvent

class StageTable(private val page: Page) : AbJTable() {
    companion object {
        private const val serialVersionUID = 1L
        private var title: Array<String>
        fun redefine() {
            title = Page.Companion.get(1, "t", 9)
        }

        init {
            redefine()
        }
    }

    protected var data: Array<Array<Any>>?
    override fun getColumnClass(c: Int): Class<*> {
        return if (c == 1) Enemy::class.java else Any::class.java
    }

    override fun getColumnCount(): Int {
        return title.size
    }

    override fun getColumnName(arg0: Int): String {
        return title[arg0]
    }

    override fun getRowCount(): Int {
        return if (data == null) 0 else data!!.size
    }

    override fun getToolTipText(e: MouseEvent): String {
        return if (columnAtPoint(e.point) == 2) {
            "{hp, atk}"
        } else {
            null
        }
    }

    override fun getValueAt(r: Int, c: Int): Any {
        if (data == null || r < 0 || c < 0 || r >= data!!.size || c >= data!![r].length) return null
        return if (c == 2) data!![r][c].toString() + "%" else data!![r][c]
    }

    fun clicked(p: Point) {
        if (data == null) return
        var c: Int = getColumnModel().getColumnIndexAtX(p.x)
        c = lnk.get(c)
        val r: Int = p.y / getRowHeight()
        if (r < 0 || r >= data!!.size || c != 1) return
        val e: Enemy = data!![r][c] as Enemy
        if (data!![r][2] !is String) return
        val d: IntArray = CommonStatic.parseIntsN(data!![r][2] as String)
        MainFrame.Companion.changePanel(EnemyInfoPage(page, e, d[0], d[1]))
    }

    fun setData(st: Stage?) {
        val info = st!!.data.simple
        data = Array(info.size) { arrayOfNulls(9) }
        for (i in info.indices) {
            val ind = info.size - i - 1
            data!![ind][1] = info[i].enemy.get()
            data!![ind][0] = if (info[i].boss == 1) "boss" else ""
            data!![ind][2] = CommonStatic.toArrayFormat(info[i].multiple, info[i].mult_atk)
            data!![ind][3] = if (info[i].number == 0) "infinite" else info[i].number
            if (info[i].castle_0 >= info[i].castle_1) data!![ind][4] = info[i].castle_0.toString() + "%" else data!![ind][4] = info[i].castle_0.toString() + "~" + info[i].castle_1 + "%"
            if (Math.abs(info[i].spawn_0) >= Math.abs(info[i].spawn_1)) data!![ind][5] = info[i].spawn_0 else data!![ind][5] = info[i].spawn_0.toString() + "~" + info[i].spawn_1
            if (info[i].respawn_0 == info[i].respawn_1) data!![ind][6] = info[i].respawn_0 else data!![ind][6] = info[i].respawn_0.toString() + "~" + info[i].respawn_1
            data!![ind][7] = if (info[i].layer_0 == info[i].layer_1) info[i].layer_0 else info[i].layer_0.toString() + "~" + info[i].layer_1
            val g = info[i].group
            val scg: SCGroup? = st.data.sub[g]
            data!![ind][8] = if (scg == null) if (g != 0) Data.Companion.trio(g) + " - invalid" else "" else scg.toString()
        }
    }

    init {
        setDefaultRenderer(Enemy::class.java, EnemyTCR())
    }
}
